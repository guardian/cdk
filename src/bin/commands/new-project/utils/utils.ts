import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
