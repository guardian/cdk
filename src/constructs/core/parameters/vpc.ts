import { GuParameter } from "./base";
import type { GuStack } from "../stack";
import type { GuNoTypeParameterProps } from "./base";

export class GuSubnetListParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "List<AWS::EC2::Subnet::Id>" });
  }
}

export class GuVpcParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::VPC::Id",
    });
  }
}
