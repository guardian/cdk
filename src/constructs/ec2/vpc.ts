import { Subnet, Vpc } from "aws-cdk-lib/aws-ec2";
import type { ISubnet, IVpc, VpcAttributes } from "aws-cdk-lib/aws-ec2";
import { GuSubnetListParameter, GuVpcParameter } from "../core";
import type { GuStack } from "../core";

interface VpcFromIdProps extends Omit<VpcAttributes, "availabilityZones"> {
  availabilityZones?: string[];
}

type VpcFromIdParameterProps = Omit<VpcFromIdProps, "vpcId">;

// TODO convert to boolean
export enum SubnetType {
  PUBLIC = "Public",
  PRIVATE = "Private",
}

// TODO: Migrate this to use `AppIdentity`
export interface GuSubnetProps {
  type?: SubnetType;
  app?: string;
}

/*
 * Some stacks maybe have multiple patterns (such as `GuEc2App`), therefore creating multiple VPC and subnet resources.
 * This function ensures that, where desired, we can have unique IDs for each VPC and subnet declaration
 */
export const maybeApp = (props?: { app?: string }): string => props?.app ?? "";

export class GuVpc {
  static subnets(scope: GuStack, subnets: string[]): ISubnet[] {
    /**
     *  As of ~1.77.0 of the aws/cdk, this line
     * [ previously Subnet.fromSubnetId(scope, `subnet-${subnetId}`, subnetId)) ]
     * resulted in the following error
     * "Found an encoded list token string in a scalar string context. Use 'Fn.select(0, list)' (not 'list[0]') to extract elements from token lists."
     * See https://github.com/aws/aws-cdk/issues/11945 for more information
     *
     * As a hacky workaround, we have moved to using the fromSubnetAttributes method and hardcoded an empty value
     * for the routeTableId prop. This prevents the error and, when tested on existing stacks, results in no change to the CFN output
     *
     * TODO: Understand VPCs and Subnets better and develop a better solution to this problem
     */
    return subnets.map((subnetId) =>
      Subnet.fromSubnetAttributes(scope, `subnet-${subnetId}`, { subnetId, routeTableId: " " })
    );
  }

  static subnetsFromParameter(scope: GuStack, props?: GuSubnetProps): ISubnet[] {
    const type = props?.type ?? SubnetType.PRIVATE;

    const subnets = new GuSubnetListParameter(scope, `${maybeApp(props)}${type}Subnets`, {
      description: `A list of ${type.toLowerCase()} subnets`,

      // TODO use `SSM_PARAMETER_PATHS.PrimaryVpcPrivateSubnets` or `SSM_PARAMETER_PATHS.PrimaryVpcPublicSubnets`
      default: `/account/vpc/primary/subnets/${type.toLowerCase()}`,
      fromSSM: true,
    });

    return GuVpc.subnets(scope, subnets.valueAsList);
  }

  static fromId(scope: GuStack, id: string, props: VpcFromIdProps): IVpc {
    return Vpc.fromVpcAttributes(scope, id, {
      availabilityZones: scope.availabilityZones,
      ...props,
    });
  }

  static fromIdParameter(scope: GuStack, id: string, props?: VpcFromIdParameterProps): IVpc {
    const vpc = GuVpcParameter.getInstance(scope);
    return GuVpc.fromId(scope, id, { ...props, vpcId: vpc.valueAsString });
  }
}
