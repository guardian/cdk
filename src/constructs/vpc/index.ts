export * from "./vpc";
export * from "./vpc-import";
