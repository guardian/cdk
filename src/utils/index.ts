export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
