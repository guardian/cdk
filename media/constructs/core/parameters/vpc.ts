import { NAMED_SSM_PARAMETER_PATHS } from "../../../constants";
import { isSingletonPresentInStack } from "../../../utils/singleton";
import type { GuStack } from "../stack";
import { GuParameter } from "./base";
import type { GuNoTypeParameterProps } from "./base";

export class GuSubnetListParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "List<AWS::EC2::Subnet::Id>" });
  }
}

/**
 * Adds a "VpcId" parameter to a stack.
 * This parameter will read from Parameter Store.
 * By default it will read from `/account/vpc/primary/id`, this can be changed at runtime if needed.
 */
export class GuVpcParameter extends GuParameter {
  private static instance: GuVpcParameter | undefined;

  private constructor(scope: GuStack) {
    super(scope, "VpcId", {
      description: NAMED_SSM_PARAMETER_PATHS.PrimaryVpcId.description,
      default: NAMED_SSM_PARAMETER_PATHS.PrimaryVpcId.path,
      type: "AWS::EC2::VPC::Id",
      fromSSM: true,
    });
  }

  public static getInstance(stack: GuStack): GuVpcParameter {
    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
      this.instance = new GuVpcParameter(stack);
    }

    return this.instance;
  }
}
