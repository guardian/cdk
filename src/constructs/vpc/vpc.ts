import type { VpcProps } from "@aws-cdk/aws-ec2";
import { GatewayVpcEndpointAwsService, SubnetType, Vpc } from "@aws-cdk/aws-ec2";
import { StringListParameter, StringParameter } from "@aws-cdk/aws-ssm";
import { VPC_SSM_PARAMETER_PREFIX } from "../../constants";
import { GuStatefulMigratableConstruct } from "../../utils/mixin";
import type { GuMigratingResource, GuStack } from "../core";

export interface GuVpcCustomProps {
  /**
   * Whether to add SSM Parameters containing VPC metadata, which are expected
   * to exist by many other Guardian CDK patterns.
   *
   * Defaults to 'true'.
   */
  ssmParameters?: boolean;

  /**
   * An identifier for the VPC to namespace SSM parameters. Customise when you
   * have multiple teams/VPCs in the same account.
   *
   * This will be combined with the /account/vpc prefix for the full parameter
   * name. e.g. '/account/vpc/primary'.
   *
   * Defaults to 'primary'.
   */
  ssmParametersNamespace?: string;
}

export interface GuVpcProps extends GuVpcCustomProps, VpcProps, GuMigratingResource {}

/**
 * Construct which creates a Virtual Private Cloud.
 *
 * NOTE: this construct requires an
 * [environment](https://docs.aws.amazon.com/cdk/latest/guide/environments.html)
 * to be set to function correctly. Without this, an environment-agnostic
 * template will be produced, which will only use two AZs even if the region
 * contains more than that.
 *
 * The VPC is provisioned with a public and private subnet for each availability
 * zone, with IPs spread evenly across these. Instances in private subnets
 * cannot be accessed directly over the internet. You should locate services in
 * instances in the private subnets, and provide HTTP(S) as required via load
 * balancers living in the public subnets.
 *
 * A CIDR block is a combined IP and network mask that specifies the IP
 * addresses in your VPC. For testing purposes any (AWS supported) CIDR block
 * will do, but for production you should:
 *
 * - use a /21 network mask (providing up to 2048 IPs, which is plenty for most
 *   accounts)
 * - ask Enterprise Tech for an IP address to go with this range (so that
 *   peering with other company VPCs is possible without IP clashes)
 *
 * A managed NAT is created for each private subnet to allow access to the
 * internet.
 *
 * For more information on VPCs and AWS see:
 * https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Subnets.html.
 */
export class GuVpc extends GuStatefulMigratableConstruct(Vpc) {
  constructor(scope: GuStack, id: string, props?: GuVpcProps) {
    const defaultVpcProps: VpcProps = {
      gatewayEndpoints: {
        s3: {
          service: GatewayVpcEndpointAwsService.S3,
        },
        dynamodb: {
          service: GatewayVpcEndpointAwsService.DYNAMODB,
        },
      },

      subnetConfiguration: [
        {
          name: "ingress",
          subnetType: SubnetType.PUBLIC,
        },
        {
          name: "application",
          subnetType: SubnetType.PRIVATE_WITH_NAT,
        },
      ],
    };

    const defaultCustomProps: Required<GuVpcCustomProps> = {
      ssmParameters: true,
      ssmParametersNamespace: "primary",
    };

    super(scope, id, { ...defaultVpcProps, ...props });

    if (props?.ssmParameters ?? defaultCustomProps.ssmParameters) {
      const namespace = props?.ssmParametersNamespace ?? defaultCustomProps.ssmParametersNamespace;
      const prefix = `${VPC_SSM_PARAMETER_PREFIX}/${namespace}`;

      new StringParameter(scope, "vpcID", {
        parameterName: `${prefix}/id`,
        stringValue: this.vpcId,
      });

      new StringListParameter(scope, "publicSubnets", {
        parameterName: `${prefix}/subnets/public`,
        stringListValue: this.publicSubnets.map((subnet) => subnet.subnetId),
      });

      new StringListParameter(scope, "privateSubnets", {
        parameterName: `${prefix}/subnets/private`,
        stringListValue: this.privateSubnets.map((subnet) => subnet.subnetId),
      });
    }
  }
}
