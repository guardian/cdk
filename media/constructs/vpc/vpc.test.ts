import { Template } from "aws-cdk-lib/assertions";
import { simpleGuStackForTesting } from "../../utils/test";
import { GuVpc } from "./vpc";

describe("The GuVpc construct", () => {
  it("should match snapshot", () => {
    const stack = simpleGuStackForTesting({
      stack: "test-stack",
      env: { region: "eu-west-1", account: "000000000000" },
    });
    new GuVpc(stack, "MyVpc");
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });
});
