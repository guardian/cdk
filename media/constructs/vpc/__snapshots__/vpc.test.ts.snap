// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuVpc construct should match snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStack",
      "GuVpc",
    ],
    "gu:cdk:version": "TEST",
  },
  "Resources": {
    "MyVpcF9F0CA6F": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "MyVpcIGW5C4A4F63": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "MyVpcVPCGW488ACE0D": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "MyVpcapplicationSubnet1DefaultRoute4B900A14": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "MyVpcingressSubnet1NATGatewayC15F3CF1",
        },
        "RouteTableId": {
          "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcapplicationSubnet1RouteTableAssociation37281D7C": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
        },
        "SubnetId": {
          "Ref": "MyVpcapplicationSubnet1Subnet0A4FCBEC",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcapplicationSubnet1RouteTableFCDD946D": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet1",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcapplicationSubnet1Subnet0A4FCBEC": {
      "Properties": {
        "AvailabilityZone": "eu-west-1a",
        "CidrBlock": "10.0.96.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "application",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet1",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcapplicationSubnet2DefaultRouteA21A20B8": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "MyVpcingressSubnet2NATGateway220B5C59",
        },
        "RouteTableId": {
          "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcapplicationSubnet2RouteTable1A5026C8": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet2",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcapplicationSubnet2RouteTableAssociationD1C39589": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
        },
        "SubnetId": {
          "Ref": "MyVpcapplicationSubnet2Subnet04E944F4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcapplicationSubnet2Subnet04E944F4": {
      "Properties": {
        "AvailabilityZone": "eu-west-1b",
        "CidrBlock": "10.0.128.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "application",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet2",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcapplicationSubnet3DefaultRouteD7A3BDF3": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "MyVpcingressSubnet3NATGateway8C3F8115",
        },
        "RouteTableId": {
          "Ref": "MyVpcapplicationSubnet3RouteTable94563292",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcapplicationSubnet3RouteTable94563292": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet3",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcapplicationSubnet3RouteTableAssociationE29CB40A": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcapplicationSubnet3RouteTable94563292",
        },
        "SubnetId": {
          "Ref": "MyVpcapplicationSubnet3Subnet36843A8F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcapplicationSubnet3Subnet36843A8F": {
      "Properties": {
        "AvailabilityZone": "eu-west-1c",
        "CidrBlock": "10.0.160.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "application",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet3",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcdynamodb4933D0A6": {
      "Properties": {
        "RouteTableIds": [
          {
            "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
          },
          {
            "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
          },
          {
            "Ref": "MyVpcapplicationSubnet3RouteTable94563292",
          },
          {
            "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
          },
          {
            "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
          },
          {
            "Ref": "MyVpcingressSubnet3RouteTable218CD36A",
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".dynamodb",
            ],
          ],
        },
        "VpcEndpointType": "Gateway",
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "MyVpcingressSubnet1DefaultRoute5C0E02BD": {
      "DependsOn": [
        "MyVpcVPCGW488ACE0D",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "RouteTableId": {
          "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcingressSubnet1EIPD447827A": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "MyVpcingressSubnet1NATGatewayC15F3CF1": {
      "DependsOn": [
        "MyVpcingressSubnet1DefaultRoute5C0E02BD",
        "MyVpcingressSubnet1RouteTableAssociationCA312C68",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "MyVpcingressSubnet1EIPD447827A",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "MyVpcingressSubnet1Subnet05B4D133",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "MyVpcingressSubnet1RouteTableAssociationCA312C68": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
        },
        "SubnetId": {
          "Ref": "MyVpcingressSubnet1Subnet05B4D133",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcingressSubnet1RouteTableBD23564A": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcingressSubnet1Subnet05B4D133": {
      "Properties": {
        "AvailabilityZone": "eu-west-1a",
        "CidrBlock": "10.0.0.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcingressSubnet2DefaultRouteF6D3CDFA": {
      "DependsOn": [
        "MyVpcVPCGW488ACE0D",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "RouteTableId": {
          "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcingressSubnet2EIP1342BA84": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "MyVpcingressSubnet2NATGateway220B5C59": {
      "DependsOn": [
        "MyVpcingressSubnet2DefaultRouteF6D3CDFA",
        "MyVpcingressSubnet2RouteTableAssociation10D8624C",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "MyVpcingressSubnet2EIP1342BA84",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "MyVpcingressSubnet2Subnet6E3FCA55",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "MyVpcingressSubnet2RouteTable17765BD0": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcingressSubnet2RouteTableAssociation10D8624C": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
        },
        "SubnetId": {
          "Ref": "MyVpcingressSubnet2Subnet6E3FCA55",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcingressSubnet2Subnet6E3FCA55": {
      "Properties": {
        "AvailabilityZone": "eu-west-1b",
        "CidrBlock": "10.0.32.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcingressSubnet3DefaultRouteEEE78E68": {
      "DependsOn": [
        "MyVpcVPCGW488ACE0D",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "RouteTableId": {
          "Ref": "MyVpcingressSubnet3RouteTable218CD36A",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcingressSubnet3EIPCC82CF52": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet3",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "MyVpcingressSubnet3NATGateway8C3F8115": {
      "DependsOn": [
        "MyVpcingressSubnet3DefaultRouteEEE78E68",
        "MyVpcingressSubnet3RouteTableAssociation0850FC5F",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "MyVpcingressSubnet3EIPCC82CF52",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "MyVpcingressSubnet3SubnetB43CCB2C",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet3",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "MyVpcingressSubnet3RouteTable218CD36A": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet3",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcingressSubnet3RouteTableAssociation0850FC5F": {
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcingressSubnet3RouteTable218CD36A",
        },
        "SubnetId": {
          "Ref": "MyVpcingressSubnet3SubnetB43CCB2C",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcingressSubnet3SubnetB43CCB2C": {
      "Properties": {
        "AvailabilityZone": "eu-west-1c",
        "CidrBlock": "10.0.64.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet3",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcs3D9E23531": {
      "Properties": {
        "RouteTableIds": [
          {
            "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
          },
          {
            "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
          },
          {
            "Ref": "MyVpcapplicationSubnet3RouteTable94563292",
          },
          {
            "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
          },
          {
            "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
          },
          {
            "Ref": "MyVpcingressSubnet3RouteTable218CD36A",
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".s3",
            ],
          ],
        },
        "VpcEndpointType": "Gateway",
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
  },
}
`;
